File: src/main/java/org/processmining/models/abstract_net_representation/Transition.java
package org.processmining.models.abstract_net_representation;

import org.processmining.models.GenerationDescription;
import org.processmining.models.Movable;

import java.util.Random;


/**
 * @author Ivan Shugurov
 *         Created  02.12.2013
 */
//T - type of place
public abstract class Transition<T extends Place> extends AbstractPetriNode implements Movable
{
    protected static final Random random = new Random();
    private T[] inputPlaces;
    private T[] outputPlaces;

    protected Transition
            (
                    org.processmining.models.graphbased.directed.petrinet.elements.Transition node,
                    GenerationDescription generationDescription,
                    T[] inPlaces, T[] outPlaces
            )
    {
        super(node, generationDescription);
        this.inputPlaces = inPlaces;
        this.outputPlaces = outPlaces;
    }

    public boolean checkAvailability()
    {
        boolean isAvailable = true;
        for (AbstractPetriNode inEdge : inputPlaces)
        {
            Place place = ((Place) inEdge);
            if (!place.hasTokens())
            {
                isAvailable = false;
                break;
            }
        }
        return isAvailable;
    }

    public T[] getOutputPlaces()  //TODO а нормально так выставлять поля?
    {
        return outputPlaces;
    }

    public T[] getInputPlaces()  //TODO а нормально так выставлять поля?
    {
        return inputPlaces;
    }

}

File: src/main/java/org/processmining/models/abstract_net_representation/Transition.kt
package org.processmining.models.abstract_net_representation

import org.processmining.models.GenerationDescription
import org.processmining.models.Movable

import java.util.Random


/**
 * @author Ivan Shugurov
 * Created  02.12.2013
 */
//P - type of place
abstract class Transition<P : Place<*>> @JvmOverloads protected constructor(
        /** Wrapped ProM transition. */
        override val node: org.processmining.models.graphbased.directed.petrinet.elements.Transition,
        generationDescription: GenerationDescription,
        val inputPlaces: List<P>,
        val outputPlaces: List<P>,
        val inputInhibitorArcPlaces: List<P> = listOf(),
        val inputResetArcPlaces: List<P> = listOf()
) : AbstractPetriNode(node, generationDescription), Movable {
    
    override fun checkAvailability() =
            (inputPlaces + inputResetArcPlaces).all { place -> place.hasTokens() } 
                    && inputInhibitorArcPlaces.all {place -> !place.hasTokens() }
    
    companion object {
        @JvmStatic
        protected val random = Random()
    }
    
}
